### ISSUE TITLE
[ap:quick] worker alpha continued after its done event (<n>:<n>:52Z): fixup+autosquash reba

### ISSUE THREAD

**Issue by WingsOfPanda** [WingsOfPanda/agglomeration-platform#204]
### Run
| | |
|---|---|
| ap version | 0.5.65 |
| command | quick |
| topic | Implement PHASES D and E ONLY (of the nine phases in section 8) of the BINDING CT-perfusion port plan at /home/liupan/ARS/production-analysis/perfusion-port-plan.md — read it fully first: section 1 (flow), section 2 (model calls), section 3 (files: the "units/*" rows of the "Backend + units" table and the whole "CTP math" table EXCEPT math/mismatch.py are your file list), section 6 (tests: boundaries A3, A4, A5, A6, A7, A8, A10 and A11 are your gates), section 7 (traps, mandatory) and section 8. Area maps: /home/liupan/ARS/production-analysis/perfusion-maps.json ("units", "math", "report" areas). Production source: /home/liupan/ARS/IFS-Perfusion (v2.2.61, HEAD c4f991f) + /home/liupan/ARS/IFS-Utils, codegraph-indexed (mcp__codegraph__codegraph_explore, projectPath=/home/liupan/ARS/IFS-Perfusion). This is run 2 of 3. Run 1 (merged, main 768b72f) delivered perfusion/backend.py (ModelBackend re-export, `_infer_labels` int64→uint8, phantom chunker), the loader rebuilt from dicom_dataloader (loader/frame.py, registry.py, axial.py, read_ctp_image) and perfusion/registration.py — read them and build on them. Run 3 does phases F (math/mismatch.py), G (reports re-baseline), H (engine rewrite, wire contract, card) and REMOVES the old TaskServe/ars_universal path. So in THIS run you again ADD modules beside the existing ones and change nothing the old engine imports: do not delete or edit construct/*, domain/*, reports/*, utils/mip.py, utils/ifs_compat.py, loader/base.py, loader/ge.py; every existing perfusion test must keep passing. PHASE D — the ten model units, each split into pure-numpy `prep`, the `backend.infer_raw` call (through irisruntime/lvm/perfusion/backend.py; model names and tensor names are measured facts in /home/liupan/ARS/production-analysis/ctp_model_io.md — input names such as `unet_2d/network_input`, `input_1_2`, `inputs/input_img`, `input_layer`; all outputs are named "output"; the four `_ArgMax` models return INT64 → `_infer_labels`), and pure-numpy `post`: units/masks.py (brain_region_seg returning BOTH masks — segment_mask and first_based_mask are distinct objects, trap 7-B; blood_supply_seg; brain_structure_seg), units/lesions.py (hemorrhage_seg with volume + ishemo; softfoci_seg with the always-zeroing gate ported VERBATIM, trap 7-C; ai_lesion_seg with the 188-line postprocess), units/aif.py (aifvof_det: histogram threshold, 384² crop, 3-slice windows, the `>=` last-wins keypoint rule, and the 537-line post with its three strategies — budget real time here, section 8 risk 4), units/vessel.py (vessel_mip_seg, vessel_atlas_seg), units/phantom.py (prep: tail cut, z-scoring, 7-feature stack, 40-phase cap; post: cbf=min(exp,50) … tmax=delay+mtt/2, NaN scatter). Do NOT contact the OM server (http://127.0.0.1:38230): every test injects the captured raw tensors. PHASE E — the CPU math core: math/filters.py (gaussian_4d_filter, mean_filter_with_mask + per-map kernel table, mean_filter, GaussianFilerWithMask with the mask argument IGNORED and in-place mutation ported verbatim, Calibrate), math/midline.py (CalculateMidLine slice pick + −49..+49° sweep, second-tier sklearn LinearRegression — scikit-learn 1.9.0 is installed; extract_mid_line_from_blood_supply; RemoveRawRadiation), math/aif_curves.py (get_mean_aif_from_blood_seg_mask — note it MUTATES both mask arguments in place, trap 7-H — filter_aif_curves, is_valid_curve, get_curve, check_structure_mask, select_a_visual_aif_point, fit_aif_to_gamma, gamma_variate, find_aif_vof_by_threshold), math/mip.py (MipEnhance, FindAIFVOFInMip — beside the old utils/mip.py, which stays until run 3), math/windows.py (PostProcess.run: percentile windows, the -1 sentinel, ×100 int16 packing, Tmax [0,1200] hard-coded, the 32767 clip only on cbv/cbf/rcbv/rcbf), math/patlak.py (CTP_St2Ct, fit_Patlak_model chunked normal equations chunk_size=65536, PS scaled 1e5/60), math/volumes.py (the volume/HIR/mismatch arithmetic as ONE function called twice in the section 7-E order; the hemo/soft-foci removal pass; remove_empty_slices). math/mismatch.py is run 3 — do not start it. GATES (section 6-A; every fixture test `@pytest.mark.live_e2e`, gated on env `ARS_CTP_STAGES_DIR`, skipping cleanly when unset, exactly like run 1's tests in irisruntime/lvm/perfusion/tests/): fixtures are /home/liupan/data/headct-golden/ctp/stages/ctp_stages.npz (17 arrays: raw_brainreg/_in, raw_aifvof_heatmap/_in, raw_lesseg/_in, phantom_prep, phantom_raw, phantom_mask, blo_sup_msk_premutation, vsl_seg_mask/_post, atl_seg_mask/_post, vsl_curves, aif_mean_list, final_registration) and ctp_stages.rest.pkl.gz (gzip pickle, pandas needed, ~16 s: `protocol` = the 37-key production debug dict incl. segment_mask, first_based_mask, brain_structure_mask, blood_supply_mask (POST-mutation), mid_line, final_registration_w_smoothing, cbf_ori…ttp_ori, result_imgs, result_mismatch, CSF_mask, softlesion_mask, aif_vof; `prepared_data`; `phantom_post`; `phantom_aif_in` (PRE-substitution AIF coords + float aif_list); `lesseg_post_in`; `vsl_final_points`; `ctp_result`; `payload`). Load the pickle ONCE per session (module-scoped fixture). Required exact tests: A3 (inject raw_brainreg → post → np.array_equal on BOTH masks vs protocol; `segment_mask is not first_based_mask`; first-wins pin), A4 (blood-supply post × first_based_mask == blo_sup_msk_premutation; label 18 present there, absent from protocol["blood_supply_mask"] and from the payload atlas), A5 (CalculateMidLine(result=None).run(final_registration_w_smoothing) == protocol["mid_line"] exactly), A6 (inject the heatmap → the `>=` argmax decode → post → the decoded cord and result_aif_vof == the PRE-substitution values in phantom_aif_in; float aif_list equality), A7 (COPIES of vsl_seg_mask/atl_seg_mask + original_image = stack(final_registration_w_smoothing)−1024 + protocol["brain_structure_mask"] + first_based_mask → aif_mean_list exact; select_a_visual_aif_point == the visual point; fit_aif_to_gamma allclose rtol=0; the mutated masks == vsl_seg_mask_post/atl_seg_mask_post), A8 (phantom.post(phantom_raw, phantom_mask) == phantom_post EXACTLY; phantom.prep(...) == phantom_prep bit-exactly incl. the tail-cut branch; clip + mean_filter_with_mask(first_based_mask) on phantom_post reproduces cbf_ori/cbv_ori/mtt_ori/tmax_ori/delay_ori), A10 (i: ai_lesion postprocess(raw_lesseg, bs, csf, mid, tmax) reproduces the AI masks — resolve the AND-with-neg_remove_mask subtlety the plan documents and say how; ii: math.volumes recompute twice in the 7-E order → volume_tmax, volume_cbf, volume_cbv, hir, mismatch_volume, mismatch_ratio EXACTLY equal to the CAPTURE run's values), A11 (windows.PostProcess.run == protocol["result_imgs"]: nine int16 maps + 18 bounds, -1 sentinel outside the mask, tmax range [0,1200], clip only on cbv/cbf/rcbv/rcbf; PS via patlak == ctp_result["PS"]). Plus the offline default-gate tests section 6 lists (test_math_filters.py, test_softfoci_is_dead.py, per-unit synthetic tests). FIXTURE-CONSISTENCY RULE (binding, overrides the plan where they differ): the plan quotes numbers from golden-headctp.json (the 14:00 run, e.g. "volume_tmax == {6: 256.018…, 10: 31.36…}" and "hir == 0.1224…"); the stage fixtures are from the 16:00 CAPTURE run. ALWAYS compare against the capture run: its payload facts are /home/liupan/data/headct-golden/ctp/golden-headctp.capture.json (AIF phase 16 / VOF 22, aif_point (5,229,260), volume_tmax {6: 262.1816578674316, 8: 119.74047191619873, 10: 28.9056862449646}) and the fixture's own `payload` / `protocol` / `ctp_result`. Never the 14:00 golden. ENVIRONMENT — your cwd is a fresh ap worktree of main (768b72f) with NO compiled cppops .so; before any gate command run `export PATH=/usr/local/Python-3.12.11/bin:$PATH; export LD_LIBRARY_PATH=/usr/local/Python-3.12.11/lib${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}; export C_INCLUDE_PATH=/usr/local/Python-3.12.11/include/python3.12; export CPLUS_INCLUDE_PATH=/usr/local/Python-3.12.11/include/python3.12; python3 setup.py build_ext --inplace`. NEVER run `pip install -e .` (any variant) in the worktree; do not pip install anything (numpy, scipy 1.15.3, scikit-image 0.26, scikit-learn 1.9.0, pandas 3.0.5, SimpleITK 2.5.6, opencv 5.0, antspyx 0.6.3 are present). Never edit the main checkout. GATE, all in the worktree, all green: (1) `PYTHON=python3 tests/run.sh` — the baseline line is CLAUDE.md:159 `**The gate baseline is "497 passed, 18 deselected"; ruff baseline is 56 errors.**`; measure and rewrite it in the same change; (2) `ruff check .` ≤ 56; (3) `python3 -c "from irisruntime.utils.cppops.cc3d import cc3d"`; (4) `bash deploy/build-wheel.sh`; (5) the fixture gates run explicitly and PASS: `ARS_CTP_STAGES_DIR=/home/liupan/data/headct-golden/ctp/stages python3 -m pytest irisruntime/lvm/perfusion/tests -m live_e2e -k "A3 or A4 or A5 or A6 or A7 or A8 or A10 or A11" -p no:cacheprovider -v`. Commit per module group (units masks+lesions / units aif+vessel+phantom / math filters+midline+aif_curves+mip / math windows+patlak+volumes). CLAUDE.md: update the "(in progress — phases A–C)" note to A–E. Final report: per-file LOC table, gate numbers, and the A3–A11 results verbatim, plus any place where the plan and the production source disagreed and which you followed (the source wins; say so). |
| run id | 11930c93-implement-phases-d-a-20260902T112458Z |
| host / user | XJP-CPU-Node / liupan |
| platform | linux · node v24.20.0 |
| providers | alpha:codex, foxtrot:claude |
| repo | git@github.com:WingsOfPanda/iris-runtime.git |
| art dir | /home/liupan/Iris/iris-runtime/.ap/state/11930c939cdc3b30daef6b51fdba41392862a5f68776713b73de93604f0a598e/implement-phases-d-a/_quick |
| filed at | 2026-09-02T11:24:58Z |

### Flag
- **hub_flag** worker alpha continued after its done event (11:21:52Z): fixup+autosquash rebase at 11:23:45Z rewrote 3 of 4 commits (30015b3/3ad0856/68411d5 -> dae78c8/0e09d75/964f2a4, AIF VOF-rise cutoff 0.3->0.5) and a gate re-run started 11:24:10Z; turn-wait had already accepted TS=ok (138s gap exceeded the 20s confirm window). Hub is holding the verify pass until the worker goes idle. _(source: from=hub command=quick)_

**Comment by WingsOfPanda** [comment]
- **part_note** the capture stores protocol.final_registration_w_smoothing after the source mutates its aliased array with -=1024. Replaying AIF prep requires +1024 and then matches raw_aifvof_heatmap_in bit-exactly; the plan describes this fixture as still HU+1024. Following source/capture and recording the disagreement. _(source: worker=alpha-codex)_
- **part_note** A6 cannot truthfully compare AIF-unit aif_list to phantom_aif_in.aif_list as requested. Source ifs.py overwrites the AIF-unit curve with vessel aif_mean_list before calling the phantom, and the capture hook records that later value; measured max difference is 297.9954. Coordinates and phases match exactly. I will gate the raw AIF unit on decoded coordinates/phases and gate phantom_aif_in.aif_list at A7, following source. _(source: worker=alpha-codex)_
- **part_note** captured phantom CBF uses scalar float32 libm exp semantics. NumPy 2.3.5 np.exp differs at 422468 voxels (max 7.63e-6); float64-rounding still differs at 700. libc expf reproduces all 1076661 captured values exactly, so the port uses that stdlib compatibility seam despite the plan calling post pure-numpy. _(source: worker=alpha-codex)_
- **part_note** ctp_stages.rest.pkl.gz stores result_mismatch only after both volume passes; the pre-AI threshold masks are not captured. A10(ii) therefore exercises the required AI-override then hemo/soft-removal calls and proves the authoritative second-pass values exactly, but the fixture cannot independently prove first-pass intermediate volumes. _(source: worker=alpha-codex)_
- **part_note** worker alpha continued after done: autosquash rebase rewrote 3 commits at 11:23:45Z (VOF-rise cutoff 0.3->0.5) and started a gate re-run; turn-1 TS=ok was accepted before the rewrite. Holding verify until the worker is idle. Budget-check: BUDGET=within ELAPSED_H=1.01 BUDGET_H=8 _(source: worker=foxtrot-claude)_

**Comment by WingsOfPanda** [comment]
- The gates were green after round 1 (517 passed, 9 fixture tests exact) and still hid nine real source-fidelity gaps; every one of them lived on a branch the single Siemens helical capture never enters (non-thalamus VOF filter, the float32 wire cast on float64 phantom maps, the soft-foci prep/gate, a skimage boundary that only bites at exactly N pixels, a 1-ulp gamma_variate re-association). For a port whose criterion is "same output as production", an exact-fixture gate is necessary but not sufficient; the line-by-line adversarial review against the production source (12 module reviewers, 3 refuters per finding) is what found them, same as run 1.
- Three of the worker's five FLAGs were plan-is-wrong calls that the judges confirmed (final_registration_w_smoothing is HU not HU+1024; A6's aif_list is unsatisfiable because ifs.py rebinds it to the vessel mean before the phantom; libm expf is the only bit-exact exp). One FLAG was wrong in the opposite direction: keeping skimage's deprecated min_size on 0.26 silently became the inclusive max_size=N, so the "parity-preserving" choice was the one that broke parity. Judge each FLAG empirically; never accept a worker's parity claim on its own.
- The worker continued past its first `done` (an autosquash rebase plus gate re-run 90s later) and turn-wait accepted TS=ok because the gap exceeded the 20s confirm window; the hub only avoided verifying a moving branch because it re-read git before starting. A confirm window measured in seconds cannot catch a worker that pauses to think; watch the worker's status/idle transition as well as the outbox.
- Two verification workflows (133 + 13 opus agents, ~11.8M tokens) cost more than the implementation turn but converted "gates green" into "faithful to source with a written list of every remaining untested branch" — the coverage-gap and plan-vs-source lists are the most valuable artifact for run 3, which wires the engine and can only be exact if these units are.
- What I would try first next time: give the module reviewers the fix-prompt shape up front (port ref, production ref, how-to-confirm) and run them BEFORE the worker's gate round, so round 1 already carries the review's findings instead of spending a second worker turn on them.

**Comment by WingsOfPanda** [comment]
triaged by /ap:review
